src/cmdline/mod.rs

```diff
@@ -336,172 +250,641 @@ impl Cmdline {
     }
 }
 
+impl UpdateArgs for BootArgs {
+    fn has_capacity(&self, more: usize) -> bool {
+        let mut auxiliary_space = 0;
+
+        if !self.0.is_empty() {
+            // this is NOT the first arg so we need to add an extra space between existing
+            // args and the new args to be added
+            auxiliary_space += 1;
+        }
+
+        self.0.len() + auxiliary_space + more <= CMDLINE_MAXSIZE
+    }
+
+    fn insert_str<T: AsRef<str>>(&mut self, slug: T) -> Result<()> {
+        let s = slug.as_ref().trim();
+        valid_str(s)?;
+
+        if !self.has_capacity(s.len()) {
+            return Err(Error::TooLarge);
+        }
+
+        if !self.0.is_empty() {
+            self.0.push(' ');
+        }
+
+        self.0.push_str(s);
+
+        Ok(())
+    }
+}
+
+/// An abstraction for the init arguments of a kernel command line.
+#[derive(Clone, Debug)]
+pub struct InitArgs(String);
+
+impl UpdateArgs for InitArgs {
+    fn has_capacity(&self, more: usize) -> bool {
+        let mut auxiliary_space = 0;
+
+        if !self.0.is_empty() {
+            // this is NOT the first arg so we need to add an extra space between existing
+            // args and the new args to be added
+            auxiliary_space += 1;
+        }
+
+        self.0.len() + auxiliary_space + more <= CMDLINE_MAXSIZE
+    }
+
+    fn insert_str<T: AsRef<str>>(&mut self, slug: T) -> Result<()> {
+        let s = slug.as_ref().trim();
+        valid_str(s)?;
+
+        if !self.has_capacity(s.len()) {
+            return Err(Error::TooLarge);
+        }
+
+        if !self.0.is_empty() {
+            self.0.push(' ');
+        }
+
+        self.0.push_str(s);
+
+        Ok(())
+    }
+}
+
+/// A builder for a kernel command line string that validates the string as it's being built.
+/// A `CString` can be constructed from this directly using `CString::new`.
+///
+/// # Examples
+///
+/// ```rust
+/// # use linux_loader::cmdline::*;
+/// # use std::ffi::CString;
+/// let mut cl = Cmdline::new(100).unwrap();
+/// cl.boot_args_mut().insert_str("foobar").unwrap();
+/// let cl_cstring = CString::new(cl.as_str().unwrap()).unwrap();
+/// assert_eq!(cl_cstring.to_str().unwrap(), "foobar");
+/// ```
+#[derive(Clone, Debug)]
+pub struct Cmdline {
+    boot_args: BootArgs,
+    init_args: InitArgs,
+    cmdline_capacity: usize,
+}
+
+impl Cmdline {
+    /// Constructs an empty [`Cmdline`] with the given capacity, including the nul terminator and ' -- ' delimiter
+    /// between boot args and init args.
+    ///
+    /// # Arguments
+    ///
+    /// * `cmdline_capacity` - Command line capacity. Must be greater than 0.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let cl = Cmdline::new(100);
+    /// ```
+    /// [`Cmdline`]: struct.Cmdline.html
+    pub fn new(cmdline_capacity: usize) -> Result<Cmdline> {
```

Comment:
With the current design of the API I don't think that a `cmdline_capacity` makes a lot of sense anymore. If we want to, we can give users the possibility to override the `MAX_CAPACITY` by having 2 contructors:
- one that takes a capacity as a parameter
- and one that doesn't take a capacity as a parameter, and just uses the defined DEFAULT instead.

Reply:
Actually, my concern here is that if we keep the capacity parameter of the Cmdline constructor and change the default MAX_CAPACITY value that we use when calling has_capacity() method of the UpdateArgs trait, we still don't handle this aspect very well: because we allow appending to the boot args string until reaching MAX_CAPACITY and also allow appending adding init args until reaches another MAX_CAPACITY in the init args string. When, eventually, call as_string() we throw an error because the joined cmdline exceeds MAX_CAPACITY; the best way to handle this case would be that in has_capacity() method to check somehow the length of both boot args string and init args string but implementing something like this means keeping a reference to init args in the BootArgs struct and vice versa which would introduce unjustified complexity from my point of view. Let's remove cmdline_capacity param from cmdline constructor and just keep a default MAX_CAPACITY for checks before inserts (preventing abuse) and before as_str as long as exceeding MAX_CAPACITY would definitely result into a erroneous cmdline; I chose 4096 as MAX_CAPACITY as long as this is the max size of a valid kernel cmdline according to the [spec](https://www.kernel.org/doc/html/v4.14/admin-guide/kernel-parameters.html)

Reply:
Ok, reading the comments below, I'll change the code and remove cmdline_capacity from constructor.